catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,396 @@
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.Validate;
 import org.codice.ddf.platform.util.StandardThreadFactoryBuilder;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
- * The SourcePoller is the scheduler of the task to poll all configured sources at a fixed interval
- * to determine their availability. It is created by the CatalogFramework's blueprint.
+ * The {@link SourcePoller} provides a non-blocking alternative to {@link Source} methods (i.e.
+ * {@link Source#isAvailable()} and {@link Source#getContentTypes()}, which can block because they
+ * submit a request to a remote client);
+ *
+ * <p>This class is instantiated by the CatalogFramework's blueprint, and {@link #pollAllSources} is
+ * scheduled by the {@link ddf.catalog.util.impl.SourcePollerRunner} to run at a configured fixed
+ * rate.
+ *
+ * <p>TODO Add some more details about known issues and about design decisions to help prevent
+ * future regression
  *
- * <p>An isAvailable() method is included in this class so that the caller, nominally the
- * CatalogFramework, can retrieve the cached availability of a specific source, or have it polled on
- * demand if there is no availability status cached.
+ * @param <T> the type of value returned by the constructor parameter {@code #getValueFunction} and
+ *     by {@link #getCachedValue(Source)}
+ * @see SourcePollerRunner
+ * @see ddf.catalog.CatalogFramework#getSourceInfo(ddf.catalog.operation.SourceInfoRequest)
  */
-public class SourcePoller {
+public class SourcePoller<T> {
+
   private static final Logger LOGGER = LoggerFactory.getLogger(SourcePoller.class);
 
-  private static final int INITIAL_DELAY = 0;
+  private static final int DEFAULT_THREAD_POOL_SIZE = 128;
+
+  private static final long MINIMUM_TIMEOUT = 1;
 
-  private int interval = 1;
+  final Function<Source, T> getValueFunction;
 
-  private ScheduledExecutorService scheduler;
+  final Optional<T> optionalValueOnTimeout;
 
-  private ScheduledFuture<?> handle;
+  final Optional<T> optionalValueOnException;
 
-  private SourcePollerRunner runner;
+  final ListeningExecutorService checkThreadPool =
+      MoreExecutors.listeningDecorator(
+          Executors.newCachedThreadPool(
+              StandardThreadFactoryBuilder.newThreadFactory("checkThread")));
+
+  final ScheduledExecutorService watcherThreadPool =
+      Executors.newScheduledThreadPool(
+          NumberUtils.toInt(
+              AccessController.doPrivileged(
+                  (PrivilegedAction<String>)
+                      () -> System.getProperty("org.codice.ddf.system.threadPoolSize")),
+              DEFAULT_THREAD_POOL_SIZE),
+          StandardThreadFactoryBuilder.newThreadFactory("watcherThread"));
+
+  final Cache<SourceKey, Optional<T>> cache =
+      CacheBuilder.newBuilder()
+          .maximumSize(1000)
+          // TODO .expireAfterWrite(long duration, TimeUnit unit)?
+          .removalListener(
```

Comment:
❗️We need to remember to increase the concurrency level of the cache since it gets updated by all the futures.

Reply:
See https://github.com/codice/ddf/pull/3858#discussion_r232428314